Sources/PackageCollections/Providers/JSONPackageCollectionProvider.swift

```diff
@@ -0,0 +1,302 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2020 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+
+import Basics
+import Dispatch
+import struct Foundation.Date
+import class Foundation.JSONDecoder
+import struct Foundation.URL
+import PackageModel
+import SourceControl
+
+struct JSONPackageCollectionProvider: PackageCollectionProvider {
+    let configuration: Configuration
+    let httpClient: HTTPClient
+    let decoder: JSONDecoder
+
+    init(configuration: Configuration = .init(), httpClient: HTTPClient = .init()) {
+        self.configuration = configuration
+        self.httpClient = httpClient
+        self.decoder = JSONDecoder()
+        #if os(Linux)
+        self.decoder.dateDecodingStrategy = .iso8601
+        #else
+        if #available(macOS 10.12, iOS 10.0, watchOS 3.0, tvOS 10.0, *) {
+            self.decoder.dateDecodingStrategy = .iso8601
+        } else {
+            self.decoder.dateDecodingStrategy = .customISO8601
+        }
+        #endif
+    }
+
+    func get(_ source: PackageCollectionsModel.CollectionSource, callback: @escaping (Result<PackageCollectionsModel.Collection, Error>) -> Void) {
+        guard case .json = source.type else {
+            preconditionFailure("JSONPackageCollectionProvider can only be used for fetching 'json' package collections")
+        }
+
+        if let errors = source.validate() {
+            return callback(.failure(MultipleErrors(errors)))
+        }
+
+        // first do a head request to check content size compared to the maximumSizeInBytes constraint
+        let headOptions = self.makeRequestOptions(validResponseCodes: [200])
+        self.httpClient.head(source.url, options: headOptions) { result in
+            switch result {
+            case .failure(let error):
+                callback(.failure(error))
+            case .success(let response):
+                if let contentLength = response.headers.get("Content-Length").first.flatMap(Int.init),
+                    contentLength >= self.configuration.maximumSizeInBytes {
+                    return callback(.failure(Errors.responseTooLarge(contentLength)))
+                }
+                // next do a get request to get the actual content
+                let getOptions = self.makeRequestOptions(validResponseCodes: [200])
+                self.httpClient.get(source.url, options: getOptions) { result in
```

Comment:
Is there a way to limit how much data to fetch if there's no `Content-Length` in the headers?

Reply:
(which is dealt with as bad actor, below, but not until after having received all the data, I think?)

Reply:
not that I am aware of. do you have an idea?

Reply:
No, I'm not familiar with these networking APIs at all (only socket level where the amount of data to read is explicitly controlled).  I was hoping there would be something in the URLSession options, but then again, if there were I'm sure you would have used them.  Seems like the kind of thing that an API would provide to avoid data flooding by bad actors.

Reply:
right, we could try and stream the body and check for the actual content length as its streamed down. this would be a good enhancement.